### ISSUE TITLE
Differences between 1.5.2 and latest version

### ISSUE THREAD

**Issue by pppierre** [Pylons/pyramid#1477]
We use Pyramid for a while to deploy our Open Source solution PubliForge. For the time beeing, it works fine with tre 1.5.2 version.
Periodically, we check our product with the latest release of Pyramid. This time, we have an issue we do not understand. Here is the trackball:

Traceback (most recent call last):
  File "/home/ppierre/bin/virtualenv/local/lib/python2.7/site-packages/waitress-0.8.9-py2.7.egg/waitress/channel.py", line 337, in service
    task.service()
  File "/home/ppierre/bin/virtualenv/local/lib/python2.7/site-packages/waitress-0.8.9-py2.7.egg/waitress/task.py", line 173, in service
    self.execute()
  File "/home/ppierre/bin/virtualenv/local/lib/python2.7/site-packages/waitress-0.8.9-py2.7.egg/waitress/task.py", line 392, in execute
    app_iter = self.channel.server.application(env, start_response)
  File "/home/ppierre/bin/pyramid/pyramid/router.py", line 242, in **call**
    response = self.invoke_subrequest(request, use_tweens=True)
  File "/home/ppierre/bin/pyramid/pyramid/router.py", line 217, in invoke_subrequest
    response = handle_request(request)
  File "/home/ppierre/Devel/PubliForge/publiforge/lib/debug.py", line 37, in htmlvalidator_tween
    response = handler(request)
  File "/home/ppierre/bin/pyramid/pyramid/tweens.py", line 21, in excview_tween
    response = handler(request)
  File "/home/ppierre/bin/pyramid/pyramid/router.py", line 163, in handle_request
    response = view_callable(context, request)
  File "/home/ppierre/bin/pyramid/pyramid/config/views.py", line 240, in _secured_view
    result = _permitted(context, request)
  File "/home/ppierre/bin/pyramid/pyramid/config/views.py", line 236, in _permitted
    principals = self.authn_policy.effective_principals(request)
  File "/home/ppierre/bin/pyramid/pyramid/authentication.py", line 130, in effective_principals
    userid = self.unauthenticated_userid(request)
  File "/home/ppierre/bin/pyramid/pyramid/authentication.py", line 650, in unauthenticated_userid
    result = self.cookie.identify(request)
  File "/home/ppierre/bin/pyramid/pyramid/authentication.py", line 905, in identify
    self.secret, cookie, remote_addr, self.hashalg)
  File "/home/ppierre/bin/pyramid/pyramid/authentication.py", line 767, in parse_ticket
    if strings_differ(expected, digest):
  File "/home/ppierre/bin/pyramid/pyramid/util.py", line 260, in strings_differ
    invalid_bits += not compare_digest(left, right)
TypeError: 'unicode' does not have the buffer interface

Is there a fundamental change in the way tween works or is it a real bug?

Thanks.

**Comment by m-martinez** [comment]
I recently upgraded to Pyramid 1.6 running on Python 2.7 and I am experiencing this issue. I'm using "check_csrf_token" to check the request header value that is set in "X-CSRF-Token". It appears that "compare_digest" is still being passed a unicode value and a string value.

**Comment by mmerickel** [comment]
Can you paste a stack trace?

**Comment by m-martinez** [comment]
Sure thing. I knew I forgot something!

```
Traceback (most recent call last):
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid_debugtoolbar/toolbar.py", line 192, in toolbar_tween
    response = _handler(request)
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid_debugtoolbar/panels/performance.py", line 57, in resource_timer_handler
    result = handler(request)
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid/tweens.py", line 48, in excview_tween
    request_iface=request_iface.combined
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid/view.py", line 541, in _call_view
    response = view_callable(context, request)
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid/config/views.py", line 286, in _authdebug_view
    return view(context, request)
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid/config/views.py", line 353, in rendered_view
    result = view(context, request)
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid/tweens.py", line 20, in excview_tween
    response = handler(request)
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid_tm/__init__.py", line 101, in tm_tween
    reraise(*exc_info)
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid_tm/__init__.py", line 83, in tm_tween
    response = handler(request)
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid/router.py", line 145, in handle_request
    view_name
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid/view.py", line 541, in _call_view
    response = view_callable(context, request)
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid/config/views.py", line 602, in __call__
    return view(context, request)
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid/config/views.py", line 328, in attr_view
    return view(context, request)
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid/config/views.py", line 304, in predicate_wrapper
    return view(context, request)
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid/config/views.py", line 286, in _authdebug_view
    return view(context, request)
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid/config/views.py", line 244, in _secured_view
    return view(context, request)
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid/config/views.py", line 353, in rendered_view
    result = view(context, request)
  File "/home/mmartinez/projects/calllog/src/occams_studies/occams_studies/views/study.py", line 470, in edit_json
    check_csrf_token(request)
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid/session.py", line 129, in check_csrf_token
    if strings_differ(request.session.get_csrf_token(), supplied_token):
  File "/home/mmartinez/.virtualenvs/py2-calllog/lib/python2.7/site-packages/pyramid/util.py", line 278, in strings_differ
    invalid_bits += not compare_digest(left, right)
TypeError: 'unicode' does not have the buffer interface
```

After a bit more troubleshooting, I noticed this issue does not occur when doing POST with a "csrf_token" parameter. Rather,  it seems to occur when using the 'X-CSRF-Token' header parameter from an XHR request.

**Comment by invisibleroads** [comment]
Versions 1.6, 1.6b3, 1.6b2, 1.6b1, 1.6a2 generate the above error in Python 2.7.

Version 1.5.8 does not generate the error.

However, I can verify that version pyramid 1.6a2 did not generate the error until after I upgraded its dependencies. It looks like something changed in the dependencies. I'll investigate further.

**Comment by mmerickel** [comment]
This will be fixed in 1.6.1 via #2299.

**Comment by invisibleroads** [comment]
Cool!
